lib/runtime/allocator/freeing_bump.go

```diff
@@ -0,0 +1,586 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package allocator
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+	"math/bits"
+
+	"github.com/ChainSafe/gossamer/lib/runtime"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+const (
+	Aligment = 8
+
+	// each pointer is prefixed with 8 bytes, which indentifies the list
+	// index to which it belongs
+	HeaderSize = 8
+
+	// The minimum possible allocation size is chosen to be 8 bytes
+	// because in that case we would have easier time to provide the
+	// guaranteed alignment of 8
+	//
+	// The maximum possible allocation size is set to 32Mib
+	//
+	// NumOrders represents the number of orders supported, this number
+	// corresponds to the number of powers between the minimum an maximum
+	// possible allocation (2^3 ... 2^25 both ends inclusive)
+	NumOrders              uint32 = 23
+	MinPossibleAllocations uint32 = 8
+	MaxPossibleAllocations uint32 = (1 << 25)
+
+	PageSize     = 65536
+	MaxWasmPages = 4 * 1024 * 1024 * 1024 / PageSize
+)
+
+var (
+	bytesAllocatedSumGauge = promauto.NewGauge(prometheus.GaugeOpts{
+		Namespace: "gossamer_allocator",
+		Name:      "bytes_allocated_sum",
+		Help:      "the sum of every allocation ever made this increases every time a new allocation is made",
+	})
+	bytesAllocatedPeakGauge = promauto.NewGauge(prometheus.GaugeOpts{
+		Namespace: "gossamer_allocator",
+		Name:      "bytes_allocated_peak",
+		Help: "the peak number of bytes ever allocated this is the maximum " +
+			"the `bytes_allocated_sum` ever reached",
+	})
+	addressSpaceUsedGague = promauto.NewGauge(prometheus.GaugeOpts{
+		Namespace: "gossamer_allocator",
+		Name:      "address_space_used",
+		Help: "the amount of address space (in bytes) used by the allocator this is calculated as " +
+			"the difference between the allocator's bumper and the heap base.",
+	})
+)
+
+var (
+	ErrInvalidOrder                 = errors.New("invalid order")
```

Comment:
do these need to be exported?

Reply:
I would say yes, since I moved the allocator from `lib/runtime` to `lib/allocator`, now the allocator methods like `Allocate` and `Deallocate` can return those errors and we can use them to assert on tests in the `runtime` pkg